14: ReactApp: Tic Tac Toe

gabrielizalo opened the pull request:
# Summary

* Created the React App "Tic Tac Toe".


# Trello Card

https://trello.com/c/CIW8rlpo/43-react-app-tic-tac-toe


# Screenshot

<img width="1421" alt="reactapp-tictactoe" src="https://user-images.githubusercontent.com/656353/47319718-ff8e3580-d614-11e8-8534-865883f31199.png">

gabrielizalo committed:
Doing "Estructura de carpetas"
- The initial setup is ready.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fix some easy ES Lint errors.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Ready the "Passing Data Through Props" step.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Ready the "Making an Interactive Component" step.
  I added the Square Constructor and now all buttons print an X after
  clicked.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Ready the "Lifting State Up" step.
  Now the board's state is the central mind and every time a square is
  clicked it's updated.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Ready the "Function Components" step.
  Now the Square class is a function component.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Ready the "Taking Turns" step.
  Now the O can be marked on board and by default X begins.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Ready the "Declaring a Winner" step.
  Now we can see who's the winner.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Ready the "Lifting State Up, Again" step.
  Now we haven an history.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Ready the "Showing the Past Moves" step.
  Apart the history now we are showing it.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Ready the "Implementing Time Travel" step.
  Now we can go thought the history of buttons clicked.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fix the code and files filling Wolox React style guide.

henryzarza commented on react/react-training/src/index.js:
```diff
@@ -1,10 +1,6 @@
 
-import './scss/index.scss';
-import registerServiceWorker from './registerServiceWorker';
+import Game from './screens/Game';
```
Change this for the alias with babel

gabrielizalo replied:
I don't understand what do you mean here.

henryzarza replied:
Remember the explication that i told you

henryzarza commented on react/react-training/src/screens/Game.css:
```diff
@@ -0,0 +1,50 @@
+body {
```
Remember that the styles should be with `.scss`

henryzarza replied:
You don't delete this file. What happen?

henryzarza commented on react/react-training/src/screens/Game.css:
```diff
@@ -0,0 +1,50 @@
+  margin: 20px;
+}
+
+ol, ul {
```
Isn't good practice to assign styles to elements globals

henryzarza commented on react/react-training/src/screens/Game.css:
```diff
@@ -0,0 +1,50 @@
+  padding-left: 30px;
+}
+
+.board-row:after {
```
This `clearfix` isn't necessary so delete it

henryzarza commented on react/react-training/src/screens/Game.css:
```diff
@@ -0,0 +1,50 @@
+}
+
+.square {
+  background: #fff;
```
Remember that the colors should go in `_colors.scss` file

henryzarza commented on react/react-training/src/screens/Game.css:
```diff
@@ -0,0 +1,50 @@
+
+.square {
+  background: #fff;
+  border: 1px solid #999;
```
The same here

henryzarza commented on react/react-training/src/screens/Game.css:
```diff
@@ -0,0 +1,50 @@
+  margin-bottom: 10px;
+}
+
+.square {
```
This styles should go in `Square` component

henryzarza commented on react/react-training/src/screens/Game.css:
```diff
@@ -0,0 +1,50 @@
+  width: 34px;
+}
+
+.square:focus {
```
To use the nesting of `sass`

henryzarza commented on react/react-training/src/screens/Game.css:
```diff
@@ -0,0 +1,50 @@
+}
+
+.kbd-navigation .square:focus {
+  background: #ddd;
```
To move this to `_colors.scss` file and change the property for `background-color`

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+import React, { Component } from 'react';
```
Change the name of the file to `index.js` because that way you can import it as `import Game from './screens/Game'`

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+import React, { Component } from 'react';
+
+import Board from './MyScreenComponent/Board';
+import './Game.css';
```
To use sass module

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+import React, { Component } from 'react';
+
+import Board from './MyScreenComponent/Board';
```
Rename this folder to `component`

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+import Board from './MyScreenComponent/Board';
+import './Game.css';
+
+function calculateWinner(squares) {
```
This is a constant so move it to `constants` folder

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+}
+
+class Game extends Component {
+  constructor(props) {
```
You do not need this constructor, you can replace it by moving the definition of the state outside

gabrielizalo replied:
I don't understand what do you mean here.

henryzarza replied:
Remember the talk

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+    };
+  }
+
+  handleClick(i) {
```
Change this for a arrow function, check this [style guide](https://github.com/airbnb/javascript/tree/master/react#methods)

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+    });
+  }
+
+  jumpTo(step) {
```
Change this for a arrow function

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+    const current = history[this.state.stepNumber];
+    const winner = calculateWinner(current.squares);
+
+    const moves = history.map((step, move) => {
```
Isn't a good practice call functions in the render. I recommend you this [documentation](https://reactjs.org/docs/handling-events.html) and this [article](https://cdb.reacttraining.com/react-inline-functions-and-performance-bdff784f5578)

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+      const desc = move ? `Go to move #${move}` : 'Go to game start';
+      return (
+        <li key={desc}>
+          <button onClick={() => this.jumpTo(move)}>{desc}</button>
```
The same here

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+    return (
+      <div className="game">
+        <div className="game-board">
+          <Board squares={current.squares} onClick={i => this.handleClick(i)} />
```
The same of the functions here

henryzarza commented on react/react-training/src/screens/Game.js:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    return (
+      <div className="game">
```
To use sass module

henryzarza commented on react/react-training/src/screens/MyScreenComponent/Board.js:
```diff
@@ -0,0 +1,40 @@
+import React, { Component } from 'react';
```
This is a component that belong to Game so create a folder component and move it and rename the file for `index.js`

henryzarza commented on react/react-training/src/screens/MyScreenComponent/Board.js:
```diff
@@ -0,0 +1,40 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+
+import Square from './Square';
```
The `Square` is a component of the `Board` then do the same as above

henryzarza commented on react/react-training/src/screens/MyScreenComponent/Board.js:
```diff
@@ -0,0 +1,40 @@
+import Square from './Square';
+
+class Board extends Component {
+  renderSquare(i) {
```
Change this for a arrow function

henryzarza commented on react/react-training/src/screens/MyScreenComponent/Board.js:
```diff
@@ -0,0 +1,40 @@
+
+class Board extends Component {
+  renderSquare(i) {
+    return <Square value={this.props.squares[i]} onClick={() => this.props.onClick(i)} />;
```
Remember the corrections of the functions in the render

henryzarza commented on react/react-training/src/screens/MyScreenComponent/Board.js:
```diff
@@ -0,0 +1,40 @@
+
+  render() {
+    return (
+      <div>
```
Change this for a `Fragment`

gabrielizalo replied:
This change is made, but I think that it's better in another way. Currently, with your recommendation, we have:

``` 
// Game:
    return (
      <div className={styles.game}>
        <div className={styles.gameBoard}>
          <Board squares={current.squares} onClick={i => this.handleClick(i)} />
        </div>

// Board: 
  render() {
    return (
      <fragment>
        <div className={styles.status}>{status}</div>
           ...
```

I think it's better the board has the div and can be work independently:

``` 
// Game:
    return (
      <div className={styles.game}>
          <Board squares={current.squares} onClick={i => this.handleClick(i)} />

// Board: 
  render() {
    return (
      <div className={styles.gameBoard}>
        <div className={styles.status}>{status}</div>
           ...
```

What do you think?

henryzarza commented on react/react-training/src/screens/MyScreenComponent/Board.js:
```diff
@@ -0,0 +1,40 @@
+  render() {
+    return (
+      <div>
+        <div className="status">{status}</div>
```
Remember to use sass module

henryzarza commented on react/react-training/src/screens/MyScreenComponent/Board.js:
```diff
@@ -0,0 +1,40 @@
+
+Board.propTypes = {
+  onClick: PropTypes.string.isRequired,
+  squares: PropTypes.string.isRequired
```
The same here

henryzarza commented on react/react-training/src/screens/MyScreenComponent/Board.js:
```diff
@@ -0,0 +1,40 @@
+}
+
+Board.propTypes = {
+  onClick: PropTypes.string.isRequired,
```
This isn't string change it

henryzarza commented on react/react-training/src/screens/MyScreenComponent/Square.js:
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
```
Remember change the name of the file for `index.js` and this is a component of `Board`, check my comment above

henryzarza commented on react/react-training/src/screens/MyScreenComponent/Square.js:
```diff
@@ -0,0 +1,17 @@
+}
+
+Square.propTypes = {
+  onClick: PropTypes.string.isRequired,
```
This isn't `string` change it

henryzarza reviewed (changes requested):
The description of the commits should be more descriptive

wfolini commented:
@gabrielizalo Apply @henryzarza comments and then assign me again!

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fixed the structure of files and folders into components. Now all
  components are named index.js, all have their own styles named
  styles.scss and we are using SASS Modules.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Created a new "colors.scss" file with all colors used. Updated all
  references.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Created a new file with constants. Updated the only reference.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Changed this <div> for a <fragment>.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Moved constants into Constants folder and used as an Alis in Babel.
- Moved colors into variables/colors.scss and referenced where used.
- Moved state setting outside constructors.
- Fixed functions PropTypes, previously as strings.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fixed Board so it includes the main div (previously in Game) with his
  own style.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fixed handleClick and jumpTo (Game) and renderSquare (Board) as arrow
  functions.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Created the new component "HistoryItem" to better readability in Game
  component.
- Fixed all inside arrow functions. They are no use anymore.

gabrielizalo commented:
All recommendations are fixed.

henryzarza commented on react/react-training/src/index.js:
```diff
@@ -1,10 +1,6 @@
 
-import './scss/index.scss';
-import registerServiceWorker from './registerServiceWorker';
+import Game from './screens/Game';
```
Remember the explication that i told you

henryzarza commented on react/react-training/src/app/screens/Game/components/Board/components/Square/index.js:
```diff
@@ -0,0 +1,26 @@
+
+class Square extends Component {
+  handleClick = () => {
+    this.props.onClick(this.props.myI);
```
Use the destructuring here

henryzarza commented on react/react-training/src/app/screens/Game/components/Board/components/Square/index.js:
```diff
@@ -0,0 +1,26 @@
+  render() {
+    return (
+      <button className={styles.square} onClick={this.handleClick}>
+        {this.props.value}
```
Use the destructuring here, check this [style guide](https://github.com/Wolox/tech-guides/blob/master/react/style-guide.md#props)

henryzarza commented on react/react-training/src/app/screens/Game/components/Board/components/Square/styles.scss:
```diff
@@ -0,0 +1,21 @@
+  height: 34px;
+  margin-right: -1px;
+  margin-top: -1px;
+  padding: 0;
```
I think that this property isn't necessary

henryzarza commented on react/react-training/src/app/screens/Game/components/Board/index.js:
```diff
@@ -0,0 +1,43 @@
+    this.props.onClick(i);
+  };
+
+  renderSquare = i => <Square value={this.props.squares[i]} myI={i} onClick={this.handleClick} />;
```
Change the name of this prop (`myI`) for something more descriptive

henryzarza commented on react/react-training/src/app/screens/Game/components/Board/index.js:
```diff
@@ -0,0 +1,43 @@
+
+Board.propTypes = {
+  onClick: PropTypes.func.isRequired,
+  squares: PropTypes.string.isRequired
```
Change this definition because isn't a `string` is a `arrayOf`

henryzarza commented on react/react-training/src/app/screens/Game/components/Board/styles.scss:
```diff
@@ -0,0 +1,13 @@
+  margin-left: 20px;
+}
+
+.row:after {
```
Delete this clearfix isn't necessary

henryzarza commented on react/react-training/src/app/screens/Game/components/HistoryItem/index.js:
```diff
@@ -0,0 +1,25 @@
+
+class HistoryItem extends Component {
+  handleClick = () => {
+    this.props.handler(this.props.move);
```
Use destructuring here

henryzarza commented on react/react-training/src/app/screens/Game/components/HistoryItem/index.js:
```diff
@@ -0,0 +1,25 @@
+  };
+
+  render() {
+    const description = this.props.move ? `Go to move #${this.props.move}` : 'Go to game start';
```
Use destructuring here

henryzarza commented on react/react-training/src/app/screens/Game/components/HistoryItem/index.js:
```diff
@@ -0,0 +1,25 @@
+  render() {
+    const description = this.props.move ? `Go to move #${this.props.move}` : 'Go to game start';
+    return (
+      <li key={this.props.key}>
```
Delete this `prop` isn't necessary

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -0,0 +1,83 @@
+import React, { Component } from 'react';
+
```
Delete this break line

gabrielizalo replied:
Ignoro este error porque Es-Lint saca error. Que estos dos imports deben estar separados.

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -0,0 +1,83 @@
+import HistoryItem from './components/HistoryItem/';
+import styles from './styles.scss';
+
+function calculateWinner(squares) {
```
Move this function a `utils` or `constants` folder

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -0,0 +1,83 @@
+    }
+    squares[i] = this.state.xIsNext ? 'X' : 'O';
+    this.setState({
+      history: history.concat([
```
Change this for `history: history.concat([{ squares }])`

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -0,0 +1,83 @@
+    const winner = calculateWinner(current.squares);
+
+    let status;
+    if (winner) {
```
You can use the conditional ternary here

henryzarza commented on react/react-training/src/constants/index.js:
```diff
@@ -0,0 +1,3 @@
+const lines = [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0, 3, 6], [1, 4, 7], [2, 5, 8], [0, 4, 8], [2, 4, 6]];
+
+export default lines;
```
Don't export as default

henryzarza commented on react/react-training/src/index.js:
```diff
@@ -1,10 +1,6 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
+import Game from '@screens/Game';
+import './scss/styles.scss';
```
Do you need import this?

henryzarza commented on react/react-training/src/scss/variables/_colors.scss:
```diff
@@ -0,0 +1,3 @@
+$alto: #ddd;
```
The colors in hex should be in uppercase for consistency amongst projects, check this [style guide](https://github.com/Wolox/tech-guides/blob/master/frontend/docs/css-style-guide.md#colors)

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- All `this.props` are now deconstructed.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Renamed `myI` variable to `squareNumber`.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Deleted unnecessary CSS rules.
- Fixed the type of a property. It's an array, not a string.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Removed unnecessary key.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Moved the function `calculateWinner` to the new `utils`.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Applying several styling guides.
- Moved `calculateWinner` function to the new Utils.

gabrielizalo commented:
All new corrections readyyyy :white_check_mark:

wfolini commented on react/react-training/.babelrc:
```diff
@@ -7,8 +7,10 @@
         "~components": "./src/app/components",
-        "~screens": "./src/app/screens",
-        "~redux": "./src/redux"
+        "~redux": "./src/redux",
```
U forget change this character

wfolini commented on react/react-training/src/app/screens/Game/components/Board/index.js:
```diff
@@ -0,0 +1,47 @@
+
+Board.propTypes = {
+  onClick: PropTypes.func.isRequired,
+  squares: PropTypes.arrayOf.isRequired
```
U always need to declare what type is in the array. `PropTypes.arrayOf(PropTypes.number.isRequired).isRequired`

wfolini reviewed (commented):
Excellent! Only two comments!

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fixed the missing parameter of property arrayOf.
- Fixed all ~ characters in Babel config file.

gabrielizalo commented:
Entonces listoooo.. Ya estamos ok.. Graciassss y quedo pendienteeeee :sunglasses:

henryzarza reviewed (approved):
Good job 👍
